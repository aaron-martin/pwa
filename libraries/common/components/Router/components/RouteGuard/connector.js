/**
 * Copyright (c) 2017, Shopgate, Inc. All rights reserved.
 *
 * This source code is licensed under the Apache 2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

import { connect } from 'react-redux';

/**
 * Maps the contents of the state to the component props.
 * @param {Object} state The current application state.
 * @return {Object} The extended component props.
 */
const mapStateToProps = state => ({
  currentRoute: state.history.pathname,
});

export default connect(mapStateToProps);
