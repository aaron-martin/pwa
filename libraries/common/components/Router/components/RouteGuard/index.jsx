/**
 * Copyright (c) 2017, Shopgate, Inc. All rights reserved.
 *
 * This source code is licensed under the Apache 2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

import React, { Component } from 'react';
import PropTypes from 'prop-types';
import connect from './connector';

/**
 * The route guard is a component that will take care of hiding or removing components
 * from the component hierarchy based on the current route the component is mounted.
 */
class RouteGuard extends Component {

  static propTypes = {
    currentRoute: PropTypes.string.isRequired,
    children: PropTypes.node,
  };

  static defaultProps = {
    children: null,
  };

  /**
   * Creates a new route guard.
   * @param {Object} props The component props.
   */
  constructor(props) {
    super(props);

    this.mountedPath = null;
  }

  /**
   * Stores the mounted path of the component.
   */
  componentWillMount() {
    this.mountedPath = this.props.currentRoute;
  }

  /**
   * Conditionally renders the guard if the mounted route is active.
   * @return {JSX}
   */
  render() {
    const { currentRoute, children } = this.props;

    /**
     * If the current path does not match the mounted path, do not render the component.
     */
    if (currentRoute !== this.mountedPath) {
      return null;
    }

    /**
     * Render the component depending on the current route.
     */
    return (
      <div>
        {children}
      </div>
    );
  }
}

export default connect(RouteGuard);
