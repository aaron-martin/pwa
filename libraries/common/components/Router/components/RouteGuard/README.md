RouteGuard
---

> **Dependencies:** <br> `Library/connectors/routeGuard`

The `RouteGuard` component takes care of conditional rendering of its children
based on the current route.

If a route is changed, any component that is wrapped with a `RouteGuard` will not
render unless the original route where it was mounted became active again.

While normally the router is already taking care of when to render components,
you might want to use a `RouteGuard` in certain cirtumstances where your
component is rendered outside of the current routes root element.

An example of when to use `RouteGuard` is when you render components inside
a `Portal` (see `react-portal`) that will be visible on top of the 
DOM hierarchy at all times.

## Usage

```
import { RouteGuard } from 'Library/components';

...

<RouteGuard>
    <Portal isOpened>
        <div>
            Something that rendered in a portal 
            but only if the route where it was
            mounted is currently active.
        </div>          
    </Portal>
</RouteGuard>
```
