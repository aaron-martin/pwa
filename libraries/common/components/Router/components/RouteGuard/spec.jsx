/**
 * Copyright (c) 2017, Shopgate, Inc. All rights reserved.
 *
 * This source code is licensed under the Apache 2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

import React from 'react';
import { mount } from 'enzyme';
import RouteGuard from './index';

// Mock the redux connect() method instead of providing a fake store.
jest.mock('./connector', () => obj => obj);

describe('<RouteGuard />', () => {
  /**
   * Creates a test component.
   * @returns {JSX} The test component.
   */
  const createComponent = () => mount((
    <RouteGuard currentRoute="initialRoute">
      <div id="TheContents" />
    </RouteGuard>
  ));

  it('should render the opened portal when on same route', () => {
    const wrapper = createComponent();

    // Do not change the route.
    wrapper.setProps({ currentRoute: 'initialRoute' });

    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find('#TheContents').length).toBe(1);
  });

  it('should not render the opened portal when on another route', () => {
    const wrapper = createComponent();

    // This is what happens when the route is changed and the guard is connected to the store.
    wrapper.setProps({ currentRoute: 'anotherRoute' });

    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find('#TheContents').length).toBe(0);
  });
});
