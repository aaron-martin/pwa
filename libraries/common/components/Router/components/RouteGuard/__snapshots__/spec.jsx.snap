// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RouteGuard /> should not render the opened portal when on another route 1`] = `
<RouteGuard
  currentRoute="anotherRoute"
/>
`;

exports[`<RouteGuard /> should render the opened portal when on same route 1`] = `
<RouteGuard
  currentRoute="initialRoute"
>
  <div>
    <div
      id="TheContents"
    />
  </div>
</RouteGuard>
`;
